examples/auction/src/lib.rs

```
@@ -0,0 +1,349 @@
+        state.auction_state = AuctionState::BidsOverWaitingForAuctionFinalization;
+        bail!(ReceiveError::BidsOverWaitingForAuctionFinalization);
+    }
+
+    let sender_address = match ctx.sender() {
+        Address::Contract(_) => bail!(ReceiveError::ContractSender),
+        Address::Account(account_address) => account_address,
+    };
+    let existing_bid = *state.bids.entry(sender_address).or_insert_with(|| Amount::zero());
+
+    let new_bid = existing_bid + amount;
+    // Ensure that the new bid exceeds the highest bid so far
+    ensure!(new_bid > state.highest_bid, ReceiveError::BidTooLow { bidded: amount, highest_bid: state.highest_bid });
+    state.highest_bid = new_bid;
+    state.bids.insert(sender_address, existing_bid + amount);
+
+    Ok(A::accept())
+}
+
+#[receive(contract = "auction", name = "finalize")]
+fn auction_finalize<A: HasActions>(
+    ctx: &impl HasReceiveContext,
+    state: &mut State,
+) -> Result<A, FinalizeError> {
+    ensure!(state.auction_state <= AuctionState::BidsOverWaitingForAuctionFinalization, FinalizeError::AuctionFinalized);
+
+    let slot_time = ctx.metadata().slot_time();
+    if slot_time <= state.expiry {
+        bail!(FinalizeError::AuctionStillActive);
+    }
+
+    let owner = ctx.owner();
+
+    ensure!(state.auction_state == AuctionState::Active, FinalizeError::AuctionFinalized);
+
+    let balance = ctx.self_balance();
+    if balance == Amount::zero() {
+        Ok(A::accept())
+    } else {
+        let mut return_action = A::simple_transfer(&owner, state.highest_bid);
+        let mut remaining_bids = BTreeMap::new();
+        // Return bids that are smaller than highest
+        for (addr, &amnt) in state.bids.iter() {
+            if amnt < state.highest_bid {
+                return_action = return_action.and_then(A::simple_transfer(addr, amnt));
+            } else {
+                state.auction_state = AuctionState::Sold(*addr);
+                remaining_bids.insert(addr, amnt);
```

Here set to `Some (addr, amnt)`

---

You could raise an error if this is already set to `Some`.

---

Alternatively just do
```rust
return_action = return_action.and_then(A::simple_transfer(owner, amnt)
```